File: src/windows/lib/videoCapture.js
const Capture = Windows.Media.Capture;
const FocusMode = Windows.Media.Devices.FocusMode;
const Promise = WinJS.Promise;

let currentVideoCapture;

function create(videoDeviceId) {

  let captureSettings = new Capture.MediaCaptureInitializationSettings();
  captureSettings.streamingCaptureMode = Capture.StreamingCaptureMode.video;
  captureSettings.videoDeviceId = videoDeviceId;

  let capture = new Windows.Media.Capture.MediaCapture();
  let displayInformation = Windows.Graphics.Display.DisplayInformation.getForCurrentView();

  return capture.initializeAsync(captureSettings).then(function () {
    adjustCaptureOrientation(displayInformation);
    displayInformation.addEventListener("orientationchanged", onOrientationChange);

    let ctrl = capture.videoDeviceController;
    let canFocus = ctrl.focusControl && ctrl.focusControl.supported;

    if (canFocus && ctrl.focusControl.configure && ctrl.focusControl.supportedFocusModes) {

      function supportsFocusMode(mode) {
        return ctrl.focusControl.supportedFocusModes.indexOf(mode).returnValue;
      }

      var focusConfig = new Windows.Media.Devices.FocusSettings();
      focusConfig.autoFocusRange = Windows.Media.Devices.AutoFocusRange.normal;
      focusConfig.disableDriverFallback = false;
      if (supportsFocusMode(FocusMode.continuous)) {
        focusConfig.mode = FocusMode.continuous;
      } else if (supportsFocusMode(FocusMode.auto)) {
        focusConfig.mode = FocusMode.auto;
      }
    }

    function onOrientationChange(e) {
      adjustCaptureOrientation(e.target);
    }

    function adjustCaptureOrientation(displayInformation) {

      let orientationDegrees = displayOrientationToDegrees(displayInformation.currentOrientation)

      if (capture.getPreviewMirroring()) {
        orientationDegrees = 360 - orientationDegrees;
      }
      capture.setPreviewRotation(degreesToCaptureRotation(orientationDegrees));
    }

    function displayOrientationToDegrees(displayOrientation) {
      switch (displayOrientation) {
        case Windows.Graphics.Display.DisplayOrientations.portrait:
          return 90;
          break;
        case Windows.Graphics.Display.DisplayOrientations.landscapeFlipped:
          return 180;
          break;
        case Windows.Graphics.Display.DisplayOrientations.portraitFlipped:
          return 270;
          break;
        case Windows.Graphics.Display.DisplayOrientations.landscape:
        default:
          return 0;
          break;
      }
    }

    function degreesToCaptureRotation(degrees) {
      switch (degrees) {
        case 0:
          return Windows.Media.Capture.VideoRotation.none;
        case 270:
          return Windows.Media.Capture.VideoRotation.clockwise270Degrees;
        case 180:
          return Windows.Media.Capture.VideoRotation.clockwise180Degrees;
        case 90:
        default:
          return Windows.Media.Capture.VideoRotation.clockwise90Degrees;
      }
    }

    let videoCapture = {
      videoDeviceId: videoDeviceId,
      url: URL.createObjectURL(capture)
    };

    videoCapture.capture = capture;

    videoCapture.canEnableLight = ctrl.torchControl && ctrl.torchControl.supported;

    videoCapture.enableLight = function () {
      if (!videoCapture.canEnableLight || !ctrl) {
        return false;
      }

      function lightEnabler(lightControl) {
        if (lightControl && lightControl.supported) {
          lightControl.enabled = true;
          if (lightControl.powerSupported) {
            lightControl.powerPercent = 90;
          }
          return true;
        }
        return false;
      }

      return lightEnabler(ctrl.torchControl);
    }

    videoCapture.disableLight = function () {
      if (!videoCapture.canEnableLight || !ctrl) {
        return;
      }

      let tc = ctrl.torchControl;

      if (tc.enabled) {
        tc.enabled = false;
      }
    }

    videoCapture.focus = function () {
      const OPERATION_IS_IN_PROGRESS = -2147024567;
      const INITIAL_FOCUS_DELAY = 200;

      if (canFocus) {
        let focusControl = capture.videoDeviceController.focusControl;
        if (focusControl.focusState !== Windows.Media.Devices.MediaCaptureFocusState.searching) {
          Promise.timeout(INITIAL_FOCUS_DELAY).done(function () {
            focusControl.focusAsync().onerror = function (error) {
              if (error.number !== OPERATION_IS_IN_PROGRESS) {
                console.error(error);
              }
            };
          });


        }
      }
    }

    videoCapture.destroy = function () {
      displayInformation.removeEventListener("orientationchanged", onOrientationChange);
      return Promise.wrap();
    }

    currentVideoCapture = videoCapture;
    return videoCapture;
  });

}

function getCameras() {
  var Devices = Windows.Devices.Enumeration;

  return Devices.DeviceInformation.findAllAsync(Devices.DeviceClass.videoCapture)
    .then(function (cameras) {

      if (!cameras || cameras.length === 0) {
        throw new Error("No cameras found");
      }

      let backCameras = cameras.filter(function (camera) {
        return camera.enclosureLocation && camera.enclosureLocation.panel === Devices.Panel.back;
      });
      let frontCameras = cameras.filter(function (camera) {
        return camera.enclosureLocation && camera.enclosureLocation.panel === Devices.Panel.front;
      });

      let camerasFound = {};

      if (backCameras.length > 0) camerasFound.back = backCameras[0];
      if (frontCameras.length > 0) camerasFound.front = frontCameras[0];

      return camerasFound;
    });
}

module.exports = {
  get: function (videoDeviceId) {
    if (currentVideoCapture) {
      return currentVideoCapture.destroy().then(function () {
        return create(videoDeviceId);
      });
    }
    return create(videoDeviceId);
  },
  getCameras: getCameras
};

File: src/windows/QRReader/VideoCapture.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.Devices.Enumeration;
using Windows.Foundation;
using Windows.Media.Capture;
using Windows.Media.Devices;
using Windows.UI.Core;

namespace QRReader
{
    public struct CameraSet
    {
        public string Back;
        public string Front;
    }

    public sealed class VideoCapture
    {
        public static IAsyncOperation<CameraSet> GetCamerasAsync()
        {
            return GetCameras().AsAsyncOperation();
        }

        private static async Task<CameraSet> GetCameras()
        {
            var cameras = await DeviceInformation.FindAllAsync(DeviceClass.VideoCapture);

            if (cameras.Count == 0)
            {
                throw new Exception("No cameras found");
            }

            var backCamera = cameras.FirstOrDefault((camera) => camera.IsEnabled && camera.EnclosureLocation.Panel == Panel.Back);
            var frontCamera = cameras.FirstOrDefault((camera) => camera.IsEnabled && camera.EnclosureLocation.Panel == Panel.Front);
            
            return new CameraSet
            {
                Back = backCamera?.Id ?? "",
                Front = frontCamera?.Id ?? ""
            };
        }

        public static IAsyncOperation<VideoCapture> CreateAsync(string cameraId)
        {
            return Create(cameraId).AsAsyncOperation();
        }

        private static async Task<VideoCapture> Create(string cameraId)
        {
            var captureSettings = new MediaCaptureInitializationSettings()
            {
                VideoDeviceId = cameraId,
                StreamingCaptureMode = StreamingCaptureMode.Video
            };

            var mediaCapture = new MediaCapture();
            await mediaCapture.InitializeAsync(captureSettings);
            var videoCapture = new VideoCapture(mediaCapture);
            await videoCapture.WatchForOrientationChange(mediaCapture);
            return videoCapture;
        }


        public MediaCapture Capture { get; private set; }
        public bool CanEnableLight { get; private set; }
        public bool CanFocus { get; private set; }
        private CoreDispatcher mainDispatcher = CoreWindow.GetForCurrentThread().Dispatcher;

        private VideoCapture(MediaCapture capture)
        {
            var cameraController = capture.VideoDeviceController;

            Capture = capture;
            CanEnableLight = cameraController.TorchControl.Supported;
            CanFocus = cameraController.FocusControl.Supported;

            if (CanFocus && cameraController.FocusControl.SupportedFocusModes.Count > 0)
            {
                var focusConfig = new FocusSettings();
                focusConfig.AutoFocusRange = AutoFocusRange.Normal;
                focusConfig.DisableDriverFallback = false;
                if (cameraController.FocusControl.SupportedFocusModes.Contains(FocusMode.Continuous))
                {
                    focusConfig.Mode = FocusMode.Continuous;
                }
                else if (cameraController.FocusControl.SupportedFocusModes.Contains(FocusMode.Auto))
                {
                    focusConfig.Mode = FocusMode.Auto;
                }
            }
        }
        
        private async Task WatchForOrientationChange(MediaCapture capture)
        {
            var camera = await DeviceInformation.CreateFromIdAsync(capture.MediaCaptureSettings.VideoDeviceId);
            var cameraRotationHelper = new CameraRotationHelper(camera.EnclosureLocation);
            OrientationChanged(cameraRotationHelper);
            cameraRotationHelper.OrientationChanged += CameraRotationHelper_OrientationChanged;
        }

        private async void CameraRotationHelper_OrientationChanged(object sender, bool e)
        {
            var helper = sender as CameraRotationHelper;
            await mainDispatcher.RunAsync(CoreDispatcherPriority.Normal, () => OrientationChanged(helper));
        }

        private void OrientationChanged(CameraRotationHelper cameraRotationHelper)
        {
            var previewOrientation = cameraRotationHelper.GetCameraPreviewOrientation();
            var videoRotation = CameraRotationHelper.ConvertSimpleOrientationToVideoRotation(previewOrientation);
            Capture.SetPreviewRotation(videoRotation);
        }

        public void EnableLight()
        {

            if (!CanEnableLight)
            {
                return;
            }

            Capture.VideoDeviceController.TorchControl.Enabled = true;
            if (Capture.VideoDeviceController.TorchControl.PowerSupported)
            {
                Capture.VideoDeviceController.TorchControl.PowerPercent = 90;
            }
        }

        public void DisableLight()
        {
            if (!CanEnableLight)
            {
                return;
            }
            
            if (Capture.VideoDeviceController.TorchControl.Enabled)
            {
                Capture.VideoDeviceController.TorchControl.Enabled = false;
            }
        }

        public void Focus()
        {
            const int INITIAL_FOCUS_DELAY = 200;
            const int OPERATION_IS_IN_PROGRESS = -2147024567;

            if (CanFocus)
            {
                var focusControl = Capture.VideoDeviceController.FocusControl;
                if (focusControl.FocusState != MediaCaptureFocusState.Searching)
                {
                    Task.Run(async () =>
                    {
                        await Task.Delay(INITIAL_FOCUS_DELAY);
                        try
                        {
                            await focusControl.FocusAsync();
                        }
                        catch (Exception exc)
                        {
                            if(exc.HResult != OPERATION_IS_IN_PROGRESS)
                            {
                                System.Diagnostics.Debug.WriteLine(exc.ToString());
                                throw exc;
                            }
                        }
                    });
                }
            }
        }

        private bool isDestroyed = false;

        public async void Destroy()
        {
            if (isDestroyed) return;
            isDestroyed = true;
            await mainDispatcher.RunAsync(CoreDispatcherPriority.Normal, async () => {
                await Capture.StopPreviewAsync();
                Capture.Dispose();
            });
        }
    }
}
